File: app/src/main/java/com/example/ama/questapp/di/AppComponent.java
package com.example.ama.questapp.di;

import com.example.ama.questapp.QuestApp;
import com.example.ama.questapp.presentation.list.di.MainQuestListComponent;

import javax.inject.Singleton;

import dagger.Component;

@Singleton
@Component(modules = {AppModule.class, DbModule.class, CurrentQuestModule.class})
public interface AppComponent {
    MainQuestListComponent.Builder mainQuestListComponent();

    void inject(QuestApp app);
}

File: app/src/main/java/com/example/ama/questapp/di/AppComponent.kt
package com.example.ama.questapp.di

import android.content.Context
import com.example.ama.questapp.QuestApp
import com.example.ama.questapp.di.modules.*
import dagger.BindsInstance
import dagger.Component
import dagger.android.AndroidInjector
import dagger.android.support.AndroidSupportInjectionModule
import javax.inject.Singleton

@Singleton
@Component(modules = [
    AndroidSupportInjectionModule::class,
    ActivityModule::class,
    FragmentModule::class,
    EngineModule::class,
    InteractorModule::class,
    RepositoryModule::class,
    ViewStateFactoryModule::class,
    DatabaseModule::class])

interface AppComponent : AndroidInjector<QuestApp> {

    @Component.Builder
    interface Builder {

        @BindsInstance
        fun app(context: Context): Builder

        fun build(): AppComponent
    }
}